packages/aws-cdk/lib/api/cloudformation-deployments.ts

```diff
@@ -338,9 +345,14 @@ export class CloudFormationDeployments {
 
     const toolkitInfo = await ToolkitInfo.lookup(resolvedEnvironment, stackSdk, options.toolkitStackName);
 
-    // Publish any assets before doing the actual deploy (do not publish any assets on import operation)
-    if (options.resourcesToImport === undefined) {
-      await this.publishStackAssets(options.stack, toolkitInfo);
+    const disableAssetPublishing = options.disableAssetPublishing ?? false;
+    const isImporting = options.resourcesToImport !== undefined;
+    // Determine whether we should publish assets. We don't publish if assets
+    // are prepublished or during an import operation.
+    const shouldPublish = !disableAssetPublishing && !isImporting;
+
+    if (shouldPublish) {
+      await this._publishStackAssets(options.stack, toolkitInfo);
     }
```

Comment:
All this is to maintain `deployStack()`'s backwards compatibility, but I wonder whether publishing assets via `deployStack()` is even necessary. Given that `cdk import` is the only other place `deployStack()` is used and that we're not publishing assets there, I can't think of any good reasons to keep the old behaviour. But, I wanted to discuss this first before making that change.

Reply:
`CloudFormationDeployments` is not exposed in v2 (though it was in v1). See https://github.com/aws/aws-cdk/issues/18211

I'd want someone else to confirm this, but from what I can tell it's up for changes.